src/gui/torrentoptionsdialog.h

```diff
@@ -26,40 +26,66 @@
  * exception statement from your version.
  */
 
-#ifndef UPDOWNRATIODIALOG_H
-#define UPDOWNRATIODIALOG_H
+#pragma once
 
 #include <QDialog>
 
+namespace BitTorrent
+{
+    class Session;
+    class TorrentHandle;
+}
+
 namespace Ui
 {
-    class UpDownRatioDialog;
+    class TorrentOptionsDialog;
 }
 
-class UpDownRatioDialog final : public QDialog
+class TorrentOptionsDialog : public QDialog
 {
     Q_OBJECT
 
 public:
-    UpDownRatioDialog(bool useDefault, qreal initialValue, qreal maxValue,
-            int initialTimeValue, int maxTimeValue,
-            QWidget *parent = nullptr);
-    ~UpDownRatioDialog();
-
-    bool useDefault() const;
-    qreal ratio() const;
-    int seedingTime() const;
+    explicit TorrentOptionsDialog(QWidget *parent, const QVector<BitTorrent::TorrentHandle *> selectedTorrents);
+    ~TorrentOptionsDialog();
 
 public slots:
     void accept() override;
 
 private slots:
+    void updateUploadSpinValue(int val);
+    void updateDownloadSpinValue(int val);
+    void updateUploadSliderValue(int val);
+    void updateDownloadSliderValue(int val);
+
     void handleRatioTypeChanged();
     void enableRatioSpin();
     void enableTimeSpin();
 
+    void disableTristateCheckboxDHT();
+    void disableTristateCheckboxPEX();
+    void disableTristateCheckboxLSD();
+
 private:
-    Ui::UpDownRatioDialog *m_ui;
+    void initDialog();
+
+    int getUploadSpeedLimit() const;
+    int getDownloadSpeedLimit() const;
+
+    bool useGlobalShareLimits() const;
+    qreal getRatio() const;
+    int getSeedingTime() const;
+
+    bool isDisableDHT() const;
+    bool isDisablePEX() const;
+    bool isDisableLSD() const;
+
+    int m_globalUploadLimit;
+    int m_globalDownloadLimit;
+
+    QVector<BitTorrent::TorrentHandle *> m_selectedTorrents;
```

Comment:
It's a bad idea to store TorrentHandle pointers. Some of them can be deleted while the dialog is being shown. You should either store torrent hashes and query torrents in `accept()` (and other places after dialog is constructed) or handle torrent deletion events to adjust this list accordingly.

Reply:
I hope i did it right (fixup commit).

Reply:
>I hope i did it right (fixup commit).

Not very. I told you:
>It's a bad idea to **store** TorrentHandle pointers.

But you can (as before) get them as an input parameter. This will save you from unnecessary lookups in the constructor. Then you can store their hashes.

Reply:
Done.